File: benchmarks/tasks/account_state_evolution/starter.go
package benchmark

import "strconv"

type AccountState interface{ isAccountState() }

type Active struct{}
type Locked struct{ FailedAttempts int64 }
type Closed struct{}

func (Active) isAccountState() {}
func (Locked) isAccountState() {}
func (Closed) isAccountState() {}

func CanSignIn(state AccountState) bool {
	switch state.(type) {
	case Active:
		return true
	case Locked:
		return false
	case Closed:
		return false
	}
	return false
}

func StatusLabel(state AccountState) string {
	switch state := state.(type) {
	case Active:
		return "active"
	case Locked:
		return "locked:" + strconv.FormatInt(state.FailedAttempts, 10)
	case Closed:
		return "closed"
	}
	return "closed"
}

File: benchmarks/tasks/account_state_evolution/starter.ts
export type AccountState =
  | { readonly kind: "active" }
  | { readonly kind: "locked"; readonly failedAttempts: bigint }
  | { readonly kind: "closed" };

export function canSignIn(state: AccountState): boolean {
  switch (state.kind) {
    case "active":
      return true;
    case "locked":
      return false;
    case "closed":
      return false;
  }
}

export function statusLabel(state: AccountState): string {
  switch (state.kind) {
    case "active":
      return "active";
    case "locked":
      return `locked:${state.failedAttempts}`;
    case "closed":
      return "closed";
  }
}
